siemens-sitop: new driver for Siemens SITOP UPS500-series UPSes

mtenberge opened the pull request:
Please consider this new driver for inclusion in NUT.

Siemens has a line of small UPSes, specially designed for use in machine control cabinets with embedded industrial PCs: they operate at 24V DC instead of AC line voltages, and are suited for DIN-rail mounting.
Machines are mostly switched off by simply interrupting the mains voltage, so with the growing popularity of embedded Linux systems in industry, I expect these UPSes to be used more and more.
There are models with lead acid batteries, but also maintenance-free models with supercaps. Some models have serial ports, others have USB interfaces.

Siemens only provides a Windows application to interface with these UPSes, but they have published the protocol in the device manual.
The protocol is not compatible with existing drivers in NUT, so I have written a new driver and corresponding documentation.
Theoretically, all models from the series (both serial and USB) should work, but as mentioned in the docs, I have only tested the driver with only 1 model variant with USB port.

zykh commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+	if (!found)
+		fatalx(EXIT_FAILURE, "No data received from the UPS");
+	
+	dstate_setinfo("ups.mfr", "Siemens");
```
I'd remove these redundant 'ups.*' items.

mtenberge replied:
I got those from the skeleton driver skel.c, so that one might needs a revision as well.

zykh commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+	}
+	
+	/* This is all we know about the charge level... */
+	dstate_setinfo("battery.charge.indication",
```
This is a blocker. If you add a new var you are supposed to:
- document it in nut-names.txt / cmdvartab
- get it approved.

mtenberge replied:
I have considered to use the existing variable battery.charge, but people (and code?) would probably assume that it contains an integer (percentage), not a string ("<85%" or ">85%").
As far as I could see, the list of existing variables does not contain a variable that suits this purpose, or does it?

zykh commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+	 * B) some models in this SITOP series have a _very_ small capacity
+	 *    (< 10sec runtime), so every second might count.
+	 */
+	poll_interval = 1;
```
hmm... I'm not so sure this is the best thing to do (without even documenting it in the man page): people will complain that the driver doesn't respect their settings.

mtenberge replied:
You are right. Instead, I added a log message if the interval is set higher than 5.

zykh commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+	/* tell the UPS to shut down, then return - DO NOT SLEEP HERE */
+	upslogx(LOG_NOTICE, "instcmd: sending command R");
+
+	/* Note: the UPS does not really like to receive data.
```
Why not moving these remarks in `instcmd()` and simply call `instcmd("shutdown.return", NULL)` here?

mtenberge replied:
Yes, better.

zykh commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+	max_polls_without_data = 2;
+	maxPollsString = getval("max_polls_without_data");
+	if (maxPollsString) {
+		if (sscanf(maxPollsString, "%d", &parsed) == 1) {
```
What about using our `str_to_uint()` (the sign is ignored)?

mtenberge replied:
If it only would have been mentioned somewhere in the Developer Guide, I would have known about it ;-)

mtenberge commented on docs/man/siemens-sitop.txt:
```diff
@@ -0,0 +1,164 @@
+This driver should work with all models in the SITOP UPS500 series,
+as long as your kernel has support for the serial port device
+(see section *USB driver* below).
+Note: this driver has only been tested with the SITOP UPS500S-2.5
```
Thanks for your review and comments.
I'm new to writing manpages, and just found out that 'AsciiDoc' is the magic word to google... ;-)
I reformatted the text in various locations.

mtenberge commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+		/* Get new data from the serial port.
+		 * No extra delay, just get the chars that were already buffered.
+		 */
+		num_received = ser_get_buf(upsfd, rx_buffer + rx_count, RX_BUFFER_SIZE - rx_count, 0, 0);
```
For normal situations it is indeed intentional, because it cannot be guaranteed that a string arrives at once. When it is split up by the transport layer, it will be re-assembled automatically by the buffer append.
It doesn't hurt if stale data is lying around in the input buffer, because it will eventually be removed by the 'shift unusable data out of the buffer' action.
For comm failure, I have added a 'flush buffer' action in the next commit, because it is very easy and readable to implement. For timeout situations, the added code complexity does not weigh up against the simplicity of shifting out the stale data, so I left this as it is.

mtenberge commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+			rx_count += num_received;
+			
+			/* parse received input data: */
+			while (rx_count >= (5 + NR_DELIMITER_CHARS)) { /* all input data are strings of 5 characters plus 1 or more delimiter characters */
```
Good point. The delimiter characters are not mentioned in the protocol description, so by not relying on it will make the driver more robust. I changed this in the next commit.

mtenberge commented on drivers/siemens-sitop.c:
```diff
@@ -0,0 +1,314 @@
+			while (rx_count >= (5 + NR_DELIMITER_CHARS)) { /* all input data are strings of 5 characters plus 1 or more delimiter characters */
+				if (strncmp(rx_buffer, "BUFRD", 5) == 0) {
+					current_ups_status.battery_alarm = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
```
Done

mtenberge commented:
One issue remaining: the naming of new drivers.
Looking in the source tree, there are only a few driver names that start with nutdrv_. Most of them are qx drivers, and they contain both dashes and underscores in their names (e.g. nutdrv_qx_voltronic-qs-hex).
Compilation results in e.g. nutdrv_qx-nutdrv_qx_voltronic-qs-hex.o. I'm not sure if that should be the desired result...
Apart from the qx-drivers, there is only nutdrv_atcl_usb (which apparently uses underscores). If this is the desired new naming convention, I will rename the siemens driver to follow this convention.
Please consider adding this information to Chapter 4 of the Developer Guide ('Creating a new driver'), as there is no use in having a naming convention if nobody knows about it...

zykh commented:
Thanks for addressing those issues.
And yes, you're right on the need to update `skel.c` and document `str_*()` functions and the new naming convention (speaking of the latter, the built drivers are `nutdrv_qx` and `nutdrv_atcl_usb`... and there are only these two simply because those are the only ones added after we decided to adopt the new convention): I'll either do that in the coming days or add issues here on GitHub so that we don't forget about that.

Two (three) things left:
- I agree with you on `battery.charge` (that's why my comment was about the things to do when adding a new var) -- I'm more for a name like `battery.charge.approximation`, though, and I'm not completely sure on what should be the format of its value, apart from the fact that it should not have the '%' explicitly expressed in the value to go along with other similar vars, but, probably, something like you did (`< 85` / `> 85`) should be just fine with me, or maybe we could use an interval (e.g. `85-100`), but I'm not so fond of it, either,
- in the manpage, I think it should be better to actually mention `pollinterval` (that is, something like: 'The serial port is polled periodically, every *pollinterval*, for new data (recommended setting: every second) ...').
- (my git complains about whitespace errors).

In my opinion this is (almost) ready, so I'm adding it to 2.7.5 milestone.

mtenberge commented:
About the charge level indication variable: I started a thread about this on the nut-upsdev list, but up to now it remains quiet there… (https://alioth-lists.debian.net/pipermail/nut-upsdev/2018-April/thread.html). For now, I removed the % characters from the indication values.

I added a separate section to the man page about polling, and fixed the line endings.

aquette commented:
@zykh and @mtenberge what's the status on this one?
I'm back on trying (again) to push 2.7.5...

zykh commented:
@aquette: code-wise, this should be ready (after a rebase).

As mentioned in the last comment, the only thing left is the publication of the (approximate) level of battery charge reported (these devices only report if they are below or above 85%) and the formalisation of the var used for that purpose and of its content (i.e. document it in `docs/nut-names.txt` and in `data/cmdvartab`): what do we want to do?

a. add a new variable for things like that?
b. or, don't use that data at all?

I'm more for the 'a' option, which is the one @mtenberge already implemented, but with a name of `battery.charge.approx` (in future, we could also consider to make it a general suffix and extend its use to other vars), instead of the implemented `battery.charge.indication`.
And, for the published content, the already implemented option (`<85` / `>85`) looks (not beautiful, but) good to me.

But I'm not good at naming things...

mtenberge commented:
I don't have a strong opinion on the variable naming, I'll leave that up to you core devs to decide.

I'll see if I can do a rebase this week.

zykh commented:
@mtenberge, thanks and sorry for the huge delay.
Since, meanwhile, no one objected, later today, I'll merge this and apply the change I suggested.
So.. last chance to shout at me...
